File: timelineview/src/main/java/com/konatsup/timelineview/CellViewAdapter.java
package com.konatsup.timelineview;


import android.annotation.SuppressLint;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.BiConsumer;

public class CellViewAdapter extends RecyclerView.Adapter<CellViewHolder> {

    private ArrayList<Cell> mCellList;

    private AdapterView.OnItemClickListener mOnItemClickListener;

    public CellViewAdapter() {
        mCellList = new ArrayList<>();
    }

    public void setCellList(ArrayList list) {
        mCellList = list;
    }

    public void addItemCount(int count) {
        mCellList.addAll(generateBlankData(count));

        notifyDataSetChanged();
    }

    /* .forEach() will perform when API level is higher than 24 */
    public void update(List<Track> trackList, final int coloumCount) {
        mCellList = new ArrayList<>(Collections.nCopies(coloumCount * trackList.size(), new Cell(Color.DKGRAY)));
        for (int i = 0; i < trackList.size(); i++) {
            final int finalI = i;
            trackList.get(i).getCompositionMap().forEach(new BiConsumer<Integer, Composition>() {
                final int startPosition = finalI * coloumCount;
                @Override
                public void accept(Integer position, Composition composition) {
                    for (int j = 0; j < composition.getDuration(); j++) {
                        Cell cell = new Cell(composition);
                        if (j == 0) cell.setStart(true);
                        if (j == composition.getDuration() - 1) cell.setEnd(true);
                        mCellList.set(startPosition + position + j, cell);
                    }
                }
            });
        }
        notifyDataSetChanged();
    }

    public void addItem(int position) {
        if (position > mCellList.size()) return;
//        mCellList.add(position, generateDummyItem());
        notifyItemInserted(mCellList.size());
    }

    /*
     * Inserting a new item at the head of the list. This uses a specialized
     * RecyclerView method, notifyItemRemoved(), to trigger any enabled item
     * animations in addition to updating the view.
     */
    public void removeItem(int position) {
        if (position >= mCellList.size()) return;

        mCellList.remove(position);
        notifyItemRemoved(position);
    }

    @Override
    public CellViewHolder onCreateViewHolder(@NonNull ViewGroup container, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(container.getContext());
        View root = inflater.inflate(R.layout.cell, container, false);

        return new CellViewHolder(root, this);
    }

    @Override
    public void onBindViewHolder(@NonNull CellViewHolder itemHolder, int position) {
        Cell item = mCellList.get(position);
        itemHolder.setCellColor(item.getColor());
        itemHolder.setImage(item.getResId());
        itemHolder.setVisibility(item.isStart(), item.isEnd());
    }

    @Override
    public int getItemCount() {
        return mCellList.size();
    }

    public void setOnItemClickListener(AdapterView.OnItemClickListener onItemClickListener) {
        mOnItemClickListener = onItemClickListener;
    }

    public void onItemHolderClick(CellViewHolder itemHolder) {
        if (mOnItemClickListener != null) {
            mOnItemClickListener.onItemClick(null, itemHolder.itemView,
                    itemHolder.getAdapterPosition(), itemHolder.getItemId());
        }
    }

    @SuppressLint("UseSparseArrays")
    public static List<Cell> generateBlankData(int count) {
        ArrayList<Cell> items = new ArrayList<>();
        for (int i = 0; i < count; i++) {
            items.add(new Cell(Color.RED));
        }
        return items;
    }

}

File: timelineview/src/main/java/com/konatsup/timelineview/CellViewAdapter.kt
package com.konatsup.timelineview


import android.annotation.SuppressLint
import android.graphics.Color
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.AdapterView

import java.util.ArrayList
import java.util.Collections

class CellViewAdapter : RecyclerView.Adapter<CellViewHolder>() {

    private var cellList: ArrayList<Cell> = ArrayList()

    private var onItemClickListener: AdapterView.OnItemClickListener? = null

    fun addItemCount(count: Int) {
        cellList.addAll(generateBlankData(count))

        notifyDataSetChanged()
    }

    fun update(trackList: List<Track>, columnCount: Int) {
        cellList = ArrayList(Collections.nCopies(columnCount * trackList.size, Cell(Color.DKGRAY)))
        for (i in trackList.indices) {
            trackList[i].compositionMap.forEach {
                val startPosition = i * columnCount
                val position = it.key
                val composition = it.value
                for (j in 0 until composition.duration) {
                    val cell = Cell(composition)
                    if (j == 0) cell.isStart = true
                    if (j == composition.duration - 1) cell.isEnd = true
                    cellList[startPosition + position + j] = cell
                }
            }
        }
        notifyDataSetChanged()
    }

    fun addItem(position: Int) {
        if (position > cellList.size) return
        notifyItemInserted(cellList.size)
    }

    /*
     * Inserting a new item at the head of the list. This uses a specialized
     * RecyclerView method, notifyItemRemoved(), to trigger any enabled item
     * animations in addition to updating the view.
     */
    fun removeItem(position: Int) {
        if (position >= cellList.size) return

        cellList.removeAt(position)
        notifyItemRemoved(position)
    }

    override fun onCreateViewHolder(container: ViewGroup, viewType: Int): CellViewHolder {
        val inflater = LayoutInflater.from(container.context)
        val root = inflater.inflate(R.layout.cell, container, false)

        return CellViewHolder(root, this)
    }

    override fun onBindViewHolder(itemHolder: CellViewHolder, position: Int) {
        val item = cellList[position]
        itemHolder.setCellColor(item.color)
        itemHolder.setImage(item.resId)
        itemHolder.setVisibility(item.isStart, item.isEnd)
    }

    override fun getItemCount(): Int {
        return cellList.size
    }

    fun setOnItemClickListener(listener: AdapterView.OnItemClickListener) {
        onItemClickListener = listener
    }

    fun onItemHolderClick(itemHolder: CellViewHolder) {
        onItemClickListener?.onItemClick(null, itemHolder.itemView,
                itemHolder.adapterPosition, itemHolder.itemId)
    }

    companion object {

        @SuppressLint("UseSparseArrays")
        fun generateBlankData(count: Int): List<Cell> {
            val items = ArrayList<Cell>()
            for (i in 0 until count) {
                items.add(Cell(Color.RED))
            }
            return items
        }
    }

}
